### PR TITLE
feat: Add priority based memory reclaim framework

### PR DESCRIPTION
**Pull request by tanjialiang**
* Adds priority base reclaiming to memory reclaim framework. The priority determines which memory pool to reclaim first on the same level. This would help to make reclaim more application logic aware. 
* Make join node reclaim priority lower than others. This is because cost of reclaiming (spilling) on join node is high compared to other nodes.

### CODE REVIEW

File: velox/exec/Task.h [hunk 1]
```diff
@@ -1000,6 +1007,8 @@ class Task : public std::enable_shared_from_this<Task> {
   // executed in a single mode throughout its lifetime
   const ExecutionMode mode_;
 
+  const int32_t priority_;
```
Comment (xiaoxmeng) [inline velox/exec/Task.h:1010]: s/priority_/memoryArbitrationPriority_/?

File: velox/exec/Task.h [hunk 2]
```diff
@@ -703,7 +707,8 @@ class Task : public std::enable_shared_from_this<Task> {
       std::shared_ptr<core::QueryCtx> queryCtx,
       ExecutionMode mode,
       ConsumerSupplier consumerSupplier,
-      std::function<void(std::exception_ptr)> onError = nullptr);
+      std::function<void(std::exception_ptr)> onError = nullptr,
+      int32_t priority = 0);
```
Comment (xiaoxmeng) [inline velox/exec/Task.h:711]: Can we move this ahead of onError? We could use macro for backward compatibility?
Shall we call this memoryArbitrationPriority (or memoryReclaimPriority) to be specific? Thanks!

File: velox/exec/MemoryReclaimer.cpp [hunk 3]
```diff
@@ -117,51 +136,66 @@ uint64_t ParallelMemoryReclaimer::reclaim(
           error(nullptr) {}
   };
 
-  std::vector<std::shared_ptr<AsyncSource<ReclaimResult>>> reclaimTasks;
-  for (const auto& candidate : candidates) {
-    if (candidate.reclaimableBytes == 0) {
-      continue;
+  uint64_t reclaimedBytes{0};
+  for (uint32_t i = 0, j = i; i < candidates.size(); i = j) {
```
Comment (xiaoxmeng) [inline velox/exec/MemoryReclaimer.cpp:140]: Why we need to handle priority in parallel memory reclaimer? thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@tanjialiang LGTM % minors. Thanks!

File: velox/exec/MemoryReclaimer.cpp (hunk 3 above)
Comment (tanjialiang) [reply velox/exec/MemoryReclaimer.cpp:140]: Right, we don't need to. They all run at once.

File: velox/exec/HashJoinBridge.h [hunk 4]
```diff
@@ -174,9 +174,10 @@ bool isLeftNullAwareJoinWithFilter(
 class HashJoinMemoryReclaimer final : public MemoryReclaimer {
  public:
   static std::unique_ptr<memory::MemoryReclaimer> create(
-      std::shared_ptr<HashJoinBridge> joinBridge) {
+      std::shared_ptr<HashJoinBridge> joinBridge,
+      int32_t priority = 0) {
```
Comment (xiaoxmeng) [inline velox/exec/HashJoinBridge.h:178]: Do we enforce a priority input later?

File: velox/exec/Task.h [hunk 5]
```diff
@@ -69,14 +69,17 @@ class Task : public std::enable_shared_from_this<Task> {
   /// thread are passed on to a separate consumer.
   /// @param onError Optional callback to receive an exception if task
   /// execution fails.
+  /// @param memoryArbitrationPriority Optional priority on task that, in a
```
Comment (xiaoxmeng) [inline velox/exec/Task.h:72]: Let's be specific for memory arbitration for now?

**Review by xiaoxmeng — APPROVED** [review]
@tanjialiang thanks for the update % minors.

File: velox/exec/Task.h (hunk 5 above)
Comment (tanjialiang) [reply velox/exec/Task.h:72]: yeah

File: velox/exec/HashJoinBridge.h (hunk 4 above)
Comment (tanjialiang) [reply velox/exec/HashJoinBridge.h:178]: Yeah we can probably enforce that later. But now we don't want to enforce as the current priority setting is on Task level.

**Pull request closed without merging**